protocols/gossipsub/src/handler.rs

```diff
@@ -203,44 +206,46 @@ where
             ) {
                 // inbound idle state
                 Some(InboundSubstreamState::WaitingInput(mut substream)) => {
-                    match substream.poll() {
-                        Ok(Async::Ready(Some(message))) => {
+                    match substream.poll_next_unpin(cx) {
+                        Poll::Ready(Some(Ok(message))) => {
                             self.inbound_substream =
                                 Some(InboundSubstreamState::WaitingInput(substream));
-                            return Ok(Async::Ready(ProtocolsHandlerEvent::Custom(message)));
+                            return Poll::Ready(ProtocolsHandlerEvent::Custom(message));
+                        }
+                        Poll::Ready(Some(Err(e))) => {
+                            // TODO: Should we really just close here? Not even log the error?
```

Comment:
Open TODO.

Reply:
I think we should log the error, then close. This should potentially even be logged as an Err, as we probably should destroy the connection. 
I had originally written this for interoperability with go and really didn't want to terminate connections. So the current version keeps running as long as we have one active (inbound or outbound) stream with the peer. If we lose the inbound stream, we probably should close the connection. What do you think?

Reply:
> If we lose the inbound stream, we probably should close the connection.

Are you suggesting to set `self.keep_alive` to `false` regardless of `self.outboundstream` being `None` or `Some`?

Reply:
I'm saying we should handle the error cases symmetrically for inbound/outbound substreams. On errors we can return `ProtocolsHandlerEvent::close()` or try to close the substream and return the error if closing the substream errors.

Currently, I have it that when we close one of them, the connection remains open unless both are closed. This favours connectivity on errors (if we can successfully close after an error). Maybe this makes it more resilient and we keep this logic?